### PR TITLE
refactor: add retry logic for flight service

### PR DESCRIPTION
**Pull request by dqhl76**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

refactor: 
1. add retry logic for `do_get`(include statitcs and fragment exchange)
2. add retry logic for `do_action`
3. add test to random kill connection between clusters to improve robust

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16097)

### CODE REVIEW

File: src/query/service/src/servers/flight/flight_client.rs [hunk 1]
```diff
@@ -225,54 +289,154 @@ impl Drop for FlightReceiver {
...
     #[async_backtrace::framed]
     pub async fn recv(&self) -> Result<Option<DataPacket>> {
-        match self.rx.recv().await {
+        let receiver = self.rx.read().clone();
+        match receiver.recv().await {
             Err(_) => Ok(None),
-            Ok(Err(error)) => Err(error),
-            Ok(Ok(message)) => Ok(Some(DataPacket::try_from(message)?)),
+            Ok(Err(error)) => {
+                self.retry().await?;
```
Comment (zhang2014) [inline src/query/service/src/servers/flight/flight_client.rs:310]: Only network errors will retry

File: src/query/service/src/servers/flight/flight_client.rs [hunk 2]
```diff
@@ -225,54 +290,159 @@ impl Drop for FlightReceiver {
...
+            rx: RwLock::new(rx),
+            notify: RwLock::new(Arc::new(WatchNotify::new())),
+            connection_info,
         }
     }
 
     #[async_backtrace::framed]
     pub async fn recv(&self) -> Result<Option<DataPacket>> {
-        match self.rx.recv().await {
+        let receiver = self.rx.read().clone();
```
Comment (zhang2014) [inline src/query/service/src/servers/flight/flight_client.rs:307]:
```suggestion
        let receiver = self.rx.read();
```

Comment (dqhl76) [reply src/query/service/src/servers/flight/flight_client.rs:307]: It will raise a compile error: we cannot hold the lock when we use await.

**Review by zhang2014 — APPROVED** [review]

**Comment by dqhl76** [comment]
Failed test seems not related to this PR's change. Try again.

**Pull request merged**